### PR TITLE
MGMT-16208: Add controller stages to assisted service

### PR DESCRIPTION
**Pull request by ori-amizur**
Add the controller stages and the timeouts handling to the assisted service.  This includes handling both for "soft" and "hard" timeouts.

Add API handling for reporting controller timeouts.

Add new event for soft finalizing stage timeouts

## List all the issues related to this PR

- [x] New Feature 
- [x] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [x] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [x] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md
/cc @filanov

### CODE REVIEW

File: internal/cluster/cluster_test.go [hunk 1]
```diff
@@ -4081,3 +4081,65 @@ var _ = Describe("ResetClusterFiles", func() {
...
+	})
+	for _, st := range finalizingStages {
+		stage := st
+		It(fmt.Sprintf("success for stage '%s'", stage), func() {
+			createCluster(models.ClusterStatusFinalizing)
+			Expect(capi.UpdateFinalizingStage(ctx, clusterID, stage)).ToNot(HaveOccurred())
+			cls, err := common.GetClusterFromDB(db, clusterID, common.SkipEagerLoading)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(cls.Progress.FinalizingStage).To(Equal(stage))
+			Expect(time.Time(cls.Progress.FinalizingStageStartedAt)).To(BeTemporally("~", time.Now(), time.Second))
```
Comment (carbonin) [inline internal/cluster/cluster_test.go:4141]: To make this less fragile you could pass the time to `UpdateFinalizingStage`.

File: internal/cluster/finalizing_stages.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	models.FinalizingStageAddingRouterCa,
+	models.FinalizingStageWaitingForOlmOperators,
+	models.FinalizingStageApplyingManifests,
+	models.FinalizingStageWaitingForOlmOperatorsCsvInitialization,
+	models.FinalizingStageWaitingForOlmOperatorsCsv,
+	models.FinalizingStageDone,
+}
+
+func convertStageToEnvVar(stage models.FinalizingStage) string {
+	return fmt.Sprintf("FINALIZING_STAGE_%s_TIMEOUT", strings.ReplaceAll(strings.ToUpper(string(stage)), " ", "_"))
```
Comment (carbonin) [inline internal/cluster/finalizing_stages.go:44]: Do we have env vars for all the stage timeouts? If not this seems like a bit of overkill.

File: internal/cluster/finalizing_stages_test.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+	"time"
+
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+	"github.com/openshift/assisted-service/models"
+	"github.com/sirupsen/logrus"
+	"github.com/thoas/go-funk"
+)
+
+var _ = DescribeTable("finalizing stage timeouts",
```
Comment (carbonin) [inline internal/cluster/finalizing_stages_test.go:15]: I can't understand what any of this is testing.

Maybe don't use `DescribeTable` here.

I'd prefer something readable and repetitive than something that is impossible to read.

File: internal/cluster/transition.go [hunk 4]
```diff
@@ -577,6 +584,11 @@ func (th *transitionHandler) PostRefreshCluster(reason string) stateswitch.PostT
 		if !ok {
 			return errors.New("PostRefreshCluster invalid argument")
 		}
+		var values []interface{}
+		for _, p := range fargs {
+			values = append(values, p(sCluster))
+		}
+		reason := fmt.Sprintf(template, values...)
```
Comment (carbonin) [inline internal/cluster/transition.go:591]: This is clever, but can we give the variables better names to make this a bit more understandable?
Maybe: something like `clusterFormatFunc` instead of `p`?

File: internal/cluster/transition.go [hunk 5]
```diff
@@ -630,6 +642,66 @@ func (th *transitionHandler) PostRefreshCluster(reason string) stateswitch.PostT
...
+
+func (th *transitionHandler) finalizingStageTimeoutMinutes(sCluster *stateCluster) int64 {
+	return int64(finalizingStageTimeout(sCluster.cluster.Progress.FinalizingStage, sCluster.cluster.MonitoredOperators, th.log).Minutes())
+}
+
+func (th *transitionHandler) FinalizingStageTimeoutMinutes(sCluster *stateCluster) interface{} {
+	return th.finalizingStageTimeoutMinutes(sCluster)
+}
+
+func (th *transitionHandler) PostRefreshFinalizingStageTimedOut(sw stateswitch.StateSwitch, args stateswitch.TransitionArgs) error {
```
Comment (carbonin) [inline internal/cluster/transition.go:657]:
```suggestion
func (th *transitionHandler) PostRefreshFinalizingStageSoftTimedOut(sw stateswitch.StateSwitch, args stateswitch.TransitionArgs) error {
```

Just to be clear that this isn't the real timeout.

File: internal/cluster/transition_test.go [hunk 6]
```diff
@@ -4693,6 +4693,122 @@ var _ = Describe("ValidationResult sort", func() {
...
+		}
+		Expect(db.Create(cls).Error).ToNot(HaveOccurred())
+		return cls
+	}
+	Context("soft timeouts disabled", func() {
+		BeforeEach(func() {
+			clusterApi = NewManager(getDefaultConfig(), common.GetTestLog().WithField("pkg", "cluster-monitor"), db, testing.GetDummyNotificationStream(ctrl),
+				mockEvents, nil, mockHostAPI, mockMetric, nil, nil, operatorsManager, nil, mockS3Api, nil, nil, nil, false)
+		})
+	})
```
Comment (carbonin) [inline internal/cluster/transition_test.go:4749]: This looks like it doesn't do anything. Please remove it.

File: swagger.yaml [hunk 7]
```diff
@@ -5672,14 +5673,8 @@ definitions:
         type: string
         format: date-time
         x-go-custom-tag: gorm:"type:timestamp with time zone"
-      node_updater_started_at:
```
Comment (carbonin) [inline swagger.yaml:5675]: Why was this removed?

**Review by carbonin — CHANGES_REQUESTED** [review]

File: internal/cluster/finalizing_stages.go (hunk 2 above)
Comment (ori-amizur) [reply internal/cluster/finalizing_stages.go:44]: Yes we do.

File: internal/cluster/finalizing_stages_test.go (hunk 3 above)
Comment (ori-amizur) [reply internal/cluster/finalizing_stages_test.go:15]: Generation of test cases by code will be always unclear whatever way to run them we choose.

File: internal/cluster/transition.go (hunk 4 above)
Comment (ori-amizur) [reply internal/cluster/transition.go:591]: OK

File: internal/cluster/transition_test.go (hunk 6 above)
Comment (ori-amizur) [reply internal/cluster/transition_test.go:4749]: OK

File: swagger.yaml (hunk 7 above)
Comment (ori-amizur) [reply swagger.yaml:5675]: I created it at the start of the epic.  Decided that I don't need it.

File: internal/cluster/transition.go (hunk 5 above)
Comment (ori-amizur) [reply internal/cluster/transition.go:657]: OK

File: internal/cluster/cluster_test.go (hunk 1 above)
Comment (ori-amizur) [reply internal/cluster/cluster_test.go:4141]: It doesn't look nice.  But I can modify the test to be less fragile.

File: internal/cluster/finalizing_stages_test.go (hunk 3 above)
Comment (ori-amizur) [reply internal/cluster/finalizing_stages_test.go:15]: Refactored the code.  Hope it is clearer now.

Comment (carbonin) [reply internal/cluster/finalizing_stages_test.go:15]: This is definitely an improvement, thanks.

I have one more suggestion, but I'll put it in a separate comment and resolve this one.

File: internal/cluster/finalizing_stages_test.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+				Entry(fmt.Sprintf("%s without environment and without operators", stage), stage, nil, "", defaultTimeout))
+			ret = append(ret,
+				Entry(fmt.Sprintf("%s with environment and without operators", stage), stage, nil, "123m", 123*time.Minute))
+		}
+
+		// Test cases for non OLM stages with operators.  Should behave the same as if operators were not provided
+		for _, stage := range nonOlmStages {
+			defaultTimeout := finalizingStageDefaultTimeout(stage, log)
+			ret = append(ret,
+				Entry(fmt.Sprintf("%s without environment and with operators", stage), stage, operators, "", defaultTimeout))
```
Comment (carbonin) [inline internal/cluster/finalizing_stages_test.go:69]: Can you update the test names to describe what is being tested and the expected result? You have the info in the comment but it would be better if it were in the test name as well.

For example this would could be `"%s without environment and with operators uses the default timeout"`

File: internal/cluster/finalizing_stages_test.go [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+			ret = append(ret,
+				Entry(fmt.Sprintf("%s with environment and with operators", stage), stage, operators, "123m", 123*time.Minute))
+		}
+
+		// Test cases for OLM stages with operators.  Should return the maximum of default timeout and the timeout specified in the operators
+		for _, stage := range olmStages {
+			ret = append(ret,
+				Entry(fmt.Sprintf("%s without environment and with operators", stage), stage, operators, "", 21*time.Hour))
+			ret = append(ret,
+				Entry(fmt.Sprintf("%s with environment and with operators", stage), stage, operators, "123m", 21*time.Hour))
```
Comment (carbonin) [inline internal/cluster/finalizing_stages_test.go:79]: Maybe a better example this would be: `"uses the operator timeout in stage %s with environment and with operators"`

File: internal/cluster/finalizing_stages_test.go [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+		// Test cases for non OLM stages with operators.  Should behave the same as if operators were not provided
+		for _, stage := range nonOlmStages {
+			defaultTimeout := finalizingStageDefaultTimeout(stage, log)
+			ret = append(ret,
+				Entry(fmt.Sprintf("%s without environment and with operators", stage), stage, operators, "", defaultTimeout))
+			ret = append(ret,
+				Entry(fmt.Sprintf("%s with environment and with operators", stage), stage, operators, "123m", 123*time.Minute))
+		}
+
+		// Test cases for OLM stages with operators.  Should return the maximum of default timeout and the timeout specified in the operators
```
Comment (carbonin) [inline internal/cluster/finalizing_stages_test.go:74]: Also maybe add a test case here where the default timeout is larger than the operator ones?

Comment (ori-amizur) [reply internal/cluster/finalizing_stages_test.go:80]: OK

File: internal/cluster/finalizing_stages_test.go (hunk 8 above)
Comment (ori-amizur) [reply internal/cluster/finalizing_stages_test.go]: OK

File: internal/cluster/finalizing_stages_test.go (hunk 9 above)
Comment (ori-amizur) [reply internal/cluster/finalizing_stages_test.go]: OK

**Review by carbonin — APPROVED** [review]
